### ISSUE TITLE
fast implementation of exp

### ISSUE THREAD

**Issue by mantepse** [sagemath/sage#34611]
<div id="comment:0"></div>

Using the recursive definition

```
exp(f) = 1 + int(f' *exp(f))
```
we can make the computation of the exponential almost the same as a single multiplication.

The drawback is that we do not benefit from the error handling in `compose`.

Depends on #34552
Depends on #34636

CC:  @tscrim

Component: **combinatorics**

Keywords: **LazyPowerSeries**

Author: **Martin Rubey**

Branch/Commit: **[`5020b9d`](https://github.com/sagemath/sagetrac-mirror/commit/5020b9decfe8a8f375102c254c28cb51ce3c7a12)**

Reviewer: **Travis Scrimshaw**

_Issue created by migration from https://trac.sagemath.org/ticket/34611_

**Comment by mantepse** [comment]
Branch: **[u/mantepse/fast_implementation_of_exp](https://github.com/sagemath/sagetrac-mirror/tree/u/mantepse/fast_implementation_of_exp)**

**Comment by mantepse** [comment]
<div id="comment:2"></div>

Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/cdea82086b729509648bd306a04346ecead00cb1"><code>cdea820</code></a></td><td><code>Merge branch 'public/rings/lazy_series_revert-34383' of https://github.com/sagemath/sagetrac-mirror into u/tscrim/derivatives_lazy_series-34413</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b7f04edc591b9cc009d2b663098f95650dde3398"><code>b7f04ed</code></a></td><td><code>Merge branch 'develop' of trac.sagemath.org:sage into t/34413/derivatives_lazy_series-34413</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/2325826800eff4c209d64fa33c09d5dfa4ce78a3"><code>2325826</code></a></td><td><code>Merge branch 'u/mantepse/derivatives_lazy_series-34413' of trac.sagemath.org:sage into t/34422/implement_functorial_composition_of_lazy_symmetric_functiosn</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ebcf5c3b4cb3d238d8957d06ee96421cca939a4b"><code>ebcf5c3</code></a></td><td><code>Merge branch 'u/mantepse/implement_functorial_composition_of_lazy_symmetric_functiosn' of trac.sagemath.org:sage into t/34423/implement_arithmetic_product_of_lazy_symmetric_functions</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/4172688e381f6b8e617bcd1a7583aec50f346307"><code>4172688</code></a></td><td><code>Merge branch 'u/mantepse/implement_arithmetic_product_of_lazy_symmetric_functions' of trac.sagemath.org:sage into t/32367/replace_lazy_power_series-32367</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/41ca99b7bfee0c11c4ece0bffa0b9230a7b4f67b"><code>41ca99b</code></a></td><td><code>Merge branch 'u/mantepse/replace_lazy_power_series-32367' of trac.sagemath.org:sage into t/34470/categories_lazy_series-34470</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fb3a5cdb6745cb78983f6bf5fc88bef93f21040a"><code>fb3a5cd</code></a></td><td><code>Merge branch 'u/mantepse/categories_lazy_series-34470' of trac.sagemath.org:sage into t/34552/lazy_series_test_suite-34552</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/855d2bf04aaebfc1696e34664b48521bffd6a79f"><code>855d2bf</code></a></td><td><code>remove unused variable</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/75c275cf989059c02bf5bd11ad7ed0ba76b3ca5f"><code>75c275c</code></a></td><td><code>add documentation and doctests for _approximate_order</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/80d1e58693a633bec4564acd048c2d3020bccfd4"><code>80d1e58</code></a></td><td><code>use that exp is holonomic</code></td></tr></table>

**Comment by mantepse** [comment]
Author: **Martin Rubey**

**Comment by mantepse** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1 +1,8 @@
+Using the recursive definition
 
+```
+exp(f) = 1 + int(f' *exp(f))
+```
+we can make the computation of the exponential almost the same as a single multiplication.
+
+The drawback is that we do not benefit from the error handling in `compose`.
``````

**Comment by mantepse** [comment]
Commit: **[`80d1e58`](https://github.com/sagemath/sagetrac-mirror/commit/80d1e58693a633bec4564acd048c2d3020bccfd4)**

**Comment by mantepse** [comment]
Changed keywords from none to **LazyPowerSeries**

**Comment by mantepse** [comment]
<div id="comment:3" align="right">comment:3</div>

Here is a comparison:

```
sage: L.<z> = LazyLaurentSeriesRing(QQ)
sage: f = L(lambda n: randint(1, 100), valuation=1)
sage: g = exp(f)
sage: %time a = g[200]
CPU times: user 86.4 ms, sys: 29 µs, total: 86.4 ms
Wall time: 86.4 ms

sage: e = L(coefficients=lambda n: 1/factorial(ZZ(n)), valuation=0)
sage: h = e(f)
sage: %time b = h[200]
CPU times: user 1.7 s, sys: 20 µs, total: 1.7 s
Wall time: 1.7 s

sage: a == b
True
```

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`80d1e58`](https://github.com/sagemath/sagetrac-mirror/commit/80d1e58693a633bec4564acd048c2d3020bccfd4)** to **[`08ca5a3`](https://github.com/sagemath/sagetrac-mirror/commit/08ca5a3b8b9dac49beacd14efa49883f197d6212)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:4"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/08ca5a3b8b9dac49beacd14efa49883f197d6212"><code>08ca5a3</code></a></td><td><code>provide fast logarithm</code></td></tr></table>

**Comment by mantepse** [comment]
Dependencies: **#34552**

**Comment by mantepse** [comment]
<div id="comment:6" align="right">comment:6</div>

There is a slight problem:

```
sage: L.<z> = LazyLaurentSeriesRing(QQ)
sage: f = L(lambda n: randint(1, 10), valuation=1)
sage: g = exp(f)
sage: %time _ = g[500]
...
RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by mantepse** [comment]
<div id="comment:7" align="right">comment:7</div>

This is not a problem if we use dense series (of course) - so there are at least three solutions:

* we could create a dense ring in `exp`,

* we could modify `Stream_uninitialized.get_coefficient` to compute coefficients in order,

* we could ignore the problem, and tell the user to use dense series.

**Comment by tscrim** [comment]
<div id="comment:8" align="right">comment:8</div>

There is also another solution: Tell the user to compute some earlier coefficients:

```
sage: L.<z> = LazyLaurentSeriesRing(QQ)
sage: f = L(lambda n: randint(1, 10), valuation=1)
sage: g = exp(f)
sage: %time _ = g[100]
CPU times: user 8.87 ms, sys: 18 µs, total: 8.88 ms
Wall time: 8.74 ms
sage: %time _ = g[200]
CPU times: user 38.1 ms, sys: 0 ns, total: 38.1 ms
Wall time: 37.8 ms
sage: %time _ = g[500]
CPU times: user 678 ms, sys: 0 ns, total: 678 ms
Wall time: 678 ms
```

However, I like option (2) the best because there is not a reasonable situation where we would not essentially end up computing all of the coefficients (ultimately in order):

```
sage: len(g._coeff_stream._cache)
501
```

**Comment by mantepse** [comment]
<div id="comment:9" align="right">comment:9</div>

A similar technique works for computing the Cauchy inverse.

**Comment by tscrim** [comment]
<div id="comment:10" align="right">comment:10</div>

We might be able to limit the number of recursion calls by being careful about the order of multiplication too.

**Comment by mantepse** [comment]
<div id="comment:11" align="right">comment:11</div>

You mean `Stream_cauchy_mul(d_self, f._coeff_stream)` vs. `Stream_cauchy_mul(f._coeff_stream, d_self)`?

All of this may also be affected by #34616.

**Comment by tscrim** [comment]
<div id="comment:12" align="right">comment:12</div>

Yes, that's right. Although since #34616 might only apply to the dense case, it might not be relevant to the issue we have with the sparse case.

**Comment by mantepse** [comment]
<div id="comment:13" align="right">comment:13</div>

What do you think about making `Stream_uninitialized` always dense.  This is essentially the same as (2) above, just with less overhead.

This could be done either here or in #34636, which is ready for review. I could then make #34636 a dependency for this ticket.  I could even do a doctest there, because the problem arises already with, for example, the implicit definition of the Catalan numbers:

```
sage: L.<z> = LazyPowerSeriesRing(ZZ); C = L.undefined(); C.define(1 + z*C^2)
sage: %time _ = C[500]
...
RecursionError: maximum recursion depth exceeded while calling a Python object

sage: L.<z> = LazyPowerSeriesRing(ZZ, sparse=True); C = L.undefined(); C.define(1 + z*C^2)
sage: %time _ = C[350]
CPU times: user 201 ms, sys: 0 ns, total: 201 ms
Wall time: 201 ms
sage: L.<z> = LazyPowerSeriesRing(ZZ, sparse=False); C = L.undefined(); C.define(1 + z*C^2)
sage: %time _ = C[350]
CPU times: user 160 ms, sys: 0 ns, total: 160 ms
Wall time: 160 ms
```

**Comment by tscrim** [comment]
<div id="comment:14" align="right">comment:14</div>

Replying to [Martin Rubey](#comment%3A13):
> What do you think about making `Stream_uninitialized` always dense.  This is essentially the same as (2) above, just with less overhead.

+1

> This could be done either here or in #34636, which is ready for review. I could then make #34636 a dependency for this ticket.

That would probably be better.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`08ca5a3`](https://github.com/sagemath/sagetrac-mirror/commit/08ca5a3b8b9dac49beacd14efa49883f197d6212)** to **[`5020b9d`](https://github.com/sagemath/sagetrac-mirror/commit/5020b9decfe8a8f375102c254c28cb51ce3c7a12)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:15"></div>

Branch pushed to git repo; I updated commit sha1. Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/3e84c90c8c9c3d74c8dd2b6b328029d74144b7bc"><code>3e84c90</code></a></td><td><code>Merge branch 'u/mantepse/categories_lazy_series-34470' of trac.sagemath.org:sage into t/34552/lazy_series_test_suite-34552</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/bd3e3ebf0d7b707b9b03ab42ee7e7c71a28df782"><code>bd3e3eb</code></a></td><td><code>make sparsity a decision of the user</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/814aa7cd53e8ca569ddfd87ba1cfa498c564531d"><code>814aa7c</code></a></td><td><code>remove Stream_cauchy_invert.get_coefficient, make sparse a mandatory argument, move _is_sparse an attribute of Stream_inexact</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7043d1cf5d937c3f947634c088e8bdc2c13e4d9f"><code>7043d1c</code></a></td><td><code>remove redundant `__init__` methods, remove finished TODOs</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e6c4cae0902abbdef92b68badad66c8833021eb3"><code>e6c4cae</code></a></td><td><code>make Stream_uninitialized always dense to avoid maximal recursion error</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/957738684943d639805ed2c727da7e3ccdc6d5f1"><code>9577386</code></a></td><td><code>implement Polynomial_generic_sparse.__floordiv__</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ec1eebddfd5f8f95e2b8a0efc26cc4179d86390b"><code>ec1eebd</code></a></td><td><code>Merge branch 'u/mantepse/__floordiv___for_sparse_polynomials' of trac.sagemath.org:sage into t/34636/make_sparsity_a_decision_of_the_user</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/de424bd78aab5b7e151316644ad460b39261734f"><code>de424bd</code></a></td><td><code>make internal rings sparse or dense if the lazy series ring is sparse or dense</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/66c1208568118bd121c3b018fc9f2843776181f2"><code>66c1208</code></a></td><td><code>Merge branch 'u/mantepse/make_sparsity_a_decision_of_the_user' of trac.sagemath.org:sage into t/34611/fast_implementation_of_exp</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/5020b9decfe8a8f375102c254c28cb51ce3c7a12"><code>5020b9d</code></a></td><td><code>adapt exp and log to new sparsity</code></td></tr></table>

**Comment by mantepse** [comment]
<div id="comment:16" align="right">comment:16</div>

A fast algorithm for inversion can be found in van der Hoeven's paper, but is not completely trivial to implement, so it should be left for another ticket.

**Comment by mantepse** [comment]
Changed dependencies from **#34552** to **#34552, #34636**

**Comment by mantepse** [comment]
<div id="comment:17" align="right">comment:17</div>

As an aside, it should be very easy to get rid of the duplicated cache in `Stream_uninitialized`: currently we have a cache for `self._target`, and create another one for `self`.  I have an experimental branch, which removes `_target`, and simply sets `self._iter` to the iterator of the target.  However, I get a strange doctest failure when there are several interlinked uninitialized streams which, additionally, have positive valuation.  I don't understand the problem yet.  However, considering #34616, we probably want a slightly more advanced cache anyway, i.e., a cache whose size grows in chunks and not just one by one.  So it may be better to consider these things together.

**Comment by tscrim** [comment]
Reviewer: **Travis Scrimshaw**

**Comment by tscrim** [comment]
<div id="comment:18" align="right">comment:18</div>

Or on a separate ticket altogether.

I think this is good to go for now as it greatly improves the performance (asymptotically; I didn't check for small values, but there is likely at worst a negligible difference).

**Comment by vbraun** [comment]
Changed branch from **[u/mantepse/fast_implementation_of_exp](https://github.com/sagemath/sagetrac-mirror/tree/u/mantepse/fast_implementation_of_exp)** to **[`5020b9d`](https://github.com/sagemath/sagetrac-mirror/commit/5020b9decfe8a8f375102c254c28cb51ce3c7a12)**
